src/binary-reader-opcnt.cc

```diff
@@ -24,100 +24,240 @@
+      stream.Writef(" %g", *GetData<float>());
+      break;
+
+    case Kind::Float64:
+      stream.Writef(" %g", *GetData<double>());
+      break;
+
+    case Kind::Uint32Uint32:
+      WriteArray<uint32_t>(
+          stream, [&stream](uint32_t value) { stream.Writef("%u", value); });
+      break;
+
+    case Kind::BlockSig:
+      WriteArray<Type>(stream, [&stream](Type type) {
+        stream.Writef("%s", GetTypeName(type));
+      });
+      break;
+
+    case Kind::BrTable: {
+      WriteArray<Index>(stream, [&stream](Index index) {
+        stream.Writef("%" PRIindex, index);
+      });
+      break;
+    }
+  }
+}
+
+bool operator==(const OpcodeInfo& lhs, const OpcodeInfo& rhs) {
+  return lhs.opcode_ == rhs.opcode_ && lhs.kind_ == rhs.kind_ &&
+         lhs.data_ == rhs.data_;
+}
+
+bool operator!=(const OpcodeInfo& lhs, const OpcodeInfo& rhs) {
+  return !(lhs == rhs);
+}
+
+bool operator<(const OpcodeInfo& lhs, const OpcodeInfo& rhs) {
+  if (lhs.opcode_ < rhs.opcode_) return true;
+  if (lhs.opcode_ > rhs.opcode_) return false;
+  if (lhs.kind_ < rhs.kind_) return true;
+  if (lhs.kind_ > rhs.kind_) return false;
+  if (lhs.data_ < rhs.data_) return true;
+  if (lhs.data_ > rhs.data_) return false;
+  return false;
+}
+
+bool operator<=(const OpcodeInfo& lhs, const OpcodeInfo& rhs) {
+  return lhs < rhs || lhs == rhs;
+}
+
+bool operator>(const OpcodeInfo& lhs, const OpcodeInfo& rhs) {
+  return !(lhs <= rhs);
+}
+
+bool operator>=(const OpcodeInfo& lhs, const OpcodeInfo& rhs) {
+  return !(lhs < rhs);
+}
+
 namespace {
 
 class BinaryReaderOpcnt : public BinaryReaderNop {
  public:
-  explicit BinaryReaderOpcnt(OpcntData* data);
+  explicit BinaryReaderOpcnt(OpcodeInfoCounts* counts);
 
   Result OnOpcode(Opcode opcode) override;
-  Result OnI32ConstExpr(uint32_t value) override;
-  Result OnGetLocalExpr(Index local_index) override;
-  Result OnSetLocalExpr(Index local_index) override;
-  Result OnTeeLocalExpr(Index local_index) override;
-  Result OnLoadExpr(Opcode opcode,
-                    uint32_t alignment_log2,
-                    Address offset) override;
-  Result OnStoreExpr(Opcode opcode,
-                     uint32_t alignment_log2,
-                     Address offset) override;
+  Result OnOpcodeBare() override;
+  Result OnOpcodeUint32(uint32_t value) override;
+  Result OnOpcodeIndex(Index value) override;
+  Result OnOpcodeUint32Uint32(uint32_t value, uint32_t value2) override;
+  Result OnOpcodeUint64(uint64_t value) override;
+  Result OnOpcodeF32(uint32_t value) override;
+  Result OnOpcodeF64(uint64_t value) override;
+  Result OnOpcodeBlockSig(Index num_types, Type* sig_types) override;
+  Result OnBrTableExpr(Index num_targets,
+                       Index* target_depths,
+                       Index default_target_depth) override;
+  Result OnEndExpr() override;
+  Result OnEndFunc() override;
 
  private:
-  OpcntData* opcnt_data;
+  template <typename... Args>
+  void Emplace(Args&&... args);
```

Comment:
Why not make this have a return type of Result, and return Ok, so that OpcodeXXX routines can just return the result of Emplace, rather than a separate return statement returning Ok?